confidential-data-hub/hub/src/bin/confidential-data-hub/server/mod.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2023 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::sync::Arc;
+
+use anyhow::Result;
+use async_trait::async_trait;
+use confidential_data_hub::{hub::Hub, DataHub};
+use lazy_static::lazy_static;
+use log::debug;
+use tokio::sync::RwLock;
+use ttrpc::{asynchronous::TtrpcContext, Code, Error, Status};
+
+use crate::{
+    api::{
+        GetResourceRequest, GetResourceResponse, KeyProviderKeyWrapProtocolInput,
+        KeyProviderKeyWrapProtocolOutput, UnsealSecretInput, UnsealSecretOutput,
+    },
+    api_ttrpc::{GetResourceService, KeyProviderService, SealedSecretService},
+    server::message::{KeyProviderInput, KeyUnwrapOutput, KeyUnwrapResults},
+};
+
+lazy_static! {
+    static ref HUB: Arc<RwLock<Option<Hub>>> = Arc::new(RwLock::new(None));
+}
+
+mod message;
+
+pub struct Server;
+
+impl Server {
+    async fn init() -> Result<()> {
+        let mut writer = HUB.write().await;
+        if writer.is_none() {
+            let hub = Hub::new().await?;
+            *writer = Some(hub);
+        }
+
+        Ok(())
+    }
+
+    pub async fn new() -> Result<Self> {
+        Self::init().await?;
+        Ok(Self)
+    }
+}
+
+#[async_trait]
+impl SealedSecretService for Server {
+    async fn unseal_secret(
+        &self,
+        _ctx: &TtrpcContext,
+        input: UnsealSecretInput,
+    ) -> ::ttrpc::Result<UnsealSecretOutput> {
+        debug!("get new UnsealSecret request");
+        let reader = HUB.read().await;
+        let reader = reader.as_ref().expect("must be initialized");
+        let plaintext = reader.unseal_secret(input.secret).await.map_err(|e| {
+            let mut status = Status::new();
+            status.set_code(Code::INTERNAL);
+            status.set_message(format!("[CDH] [ERROR]: Unseal Secret failed: {e}"));
+            Error::RpcStatus(status)
+        })?;
+
+        let mut reply = UnsealSecretOutput::new();
+        reply.plaintext = plaintext;
+        debug!("send back plaintext of the sealed secret");
+        Ok(reply)
+    }
+}
+
+#[async_trait]
+impl GetResourceService for Server {
+    async fn get_resource(
+        &self,
+        _ctx: &TtrpcContext,
+        req: GetResourceRequest,
+    ) -> ::ttrpc::Result<GetResourceResponse> {
+        debug!("get new GetResource request");
+        let reader = HUB.read().await;
+        let reader = reader.as_ref().expect("must be initialized");
+        let resource = reader.get_resource(req.ResourcePath).await.map_err(|e| {
+            let mut status = Status::new();
+            status.set_code(Code::INTERNAL);
+            status.set_message(format!("[CDH] [ERROR]: Get Resource failed: {e}"));
+            Error::RpcStatus(status)
+        })?;
+
+        let mut reply = GetResourceResponse::new();
+        reply.Resource = resource;
+        debug!("send back the resource");
+        Ok(reply)
+    }
+}
+
+#[async_trait]
+impl KeyProviderService for Server {
+    async fn un_wrap_key(
```

Comment:
It always annoyed me that this wasn't `unwrap_key` in the AA but I guess it is auto-generated.

Reply:
same as mentioned. Coinincidentally, when I coded this I also thought it should be